### PR TITLE
feat(security): Implementation for setting up agent token

### PR DESCRIPTION
**Pull request by jim-wang-intel**
Part of Securing Consul Phase 1 stories, add new implementation and changes to `security-bootstrapper` setupRegistryACL:
 - Add Create agent token logic with using global policy
 - Add Agent checking logic to determine whether we should set agent token every time on agent restarting
 - Add re-use created agent token logic in the case of re-start and re-set scenarios
 - Add set agent token implementation
 - Add re-set agent token logic in the case of re-starting and non-persistent token case if needed

Closes: #3157

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?

N/A this is new feature

## What is the new behavior?
Now Consul is set with a agent token

## Does this PR introduce a breaking change?

- [ ] Yes
- [x ] No

## New Imports

- [ ] Yes
- [x ] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

## Other information
To test this locally, git clone this PR and build the dev. version of `security-bootstrapper`:
`make docker_security_bootstrapper `
(make sure to clean up your local volume or image cache first, in case it still uses cached docker image), 
and then use compose-builder to run docker stack in dev mode: `make run dev`
Observe the docker log message on `edgex-core-consul` and you should see successfully set up message like the following:

```
......
level=INFO ts=2021-03-11T22:11:05.372083311Z app=edgex-security-bootstrapper source=command.go:101 msg="Security bootstrapper running setupRegistryACL"
level=INFO ts=2021-03-11T22:11:05.374514548Z app=edgex-security-bootstrapper source=command.go:329 msg="leader [\"192.168.160.5:8300\"] is elected"
level=INFO ts=2021-03-11T22:11:05.374604792Z app=edgex-security-bootstrapper source=command.go:285 msg="found Consul leader to set up ACL"
    2021-03-11T22:11:05.375Z [WARN]  agent.server.acl: failed to remove bootstrap file: error="remove /consul/data/acl-bootstrap-reset: no such file or directory"
    2021-03-11T22:11:05.378Z [INFO]  agent.server.acl: ACL bootstrap completed
level=INFO ts=2021-03-11T22:11:05.379855616Z app=edgex-security-bootstrapper source=command.go:149 msg="successfully bootstrap registry ACL"
level=INFO ts=2021-03-11T22:11:05.380457374Z app=edgex-security-bootstrapper source=aclbootstrap.go:109 msg="bootstrap token is written to /tmp/edgex/secrets/consul-acl-token/bootstrap_token.json"
level=INFO ts=2021-03-11T22:11:05.380849652Z app=edgex-security-bootstrapper source=command.go:362 msg="successfully retrieved secretstore management token from /tmp/edgex/secrets/edgex-consul/admin/token.json"
level=INFO ts=2021-03-11T22:11:05.380912348Z app=edgex-security-bootstrapper source=command.go:162 msg="successfully get secretstore token and configuring the registry access for secretestore"
level=INFO ts=2021-03-11T22:11:05.39776416Z app=edgex-security-bootstrapper source=command.go:414 msg="successfully configure Consul access for secretstore"
level=INFO ts=2021-03-11T22:11:05.415093814Z app=edgex-security-bootstrapper source=acltokens.go:148 msg="internal Consul ACL system is ready"
level=INFO ts=2021-03-11T22:11:05.419369993Z app=edgex-security-bootstrapper source=acltokens.go:345 msg="successfully created a new agent token"
    2021-03-11T22:11:05.423Z [INFO]  agent: Updated agent's ACL token: token=agent
level=INFO ts=2021-03-11T22:11:05.423787697Z app=edgex-security-bootstrapper source=acltokens.go:401 msg="agent token is set with type [agent]"
level=INFO ts=2021-03-11T22:11:05.423888036Z app=edgex-security-bootstrapper source=command.go:221 msg="successfully set up agent token into agent"
level=INFO ts=2021-03-11T22:11:05.423934179Z app=edgex-security-bootstrapper source=command.go:179 msg="setupRegistryACL successfully done"
...............
```

### CODE REVIEW

File: internal/security/bootstrapper/command/setupacl/command.go [hunk 1]
```diff
@@ -174,6 +188,81 @@ func (c *cmd) GetCommandName() string {
 	return CommandName
 }
 
+// setupAgentToken is to set up the agent token to the running agent if haven't set up yet
```
Comment (lenny-intel) [inline internal/security/bootstrapper/command/setupacl/command.go:191]: Comment that side effect is `BootStrapACLTokenInfo` is set if passed in as nil

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+	// RUD: Read Update Delete
+	consulTokenRUDAPI = "/v1/acl/token/%s"
+)
+
+// isACLTokenPersistent checks Consul agent's configuration property for EnablePersistence of ACLTokens
+// it returns true if the token persistence is enabled; false otherwise
+// once ACL rules are enforced, this call requires at least agent read permission and hence we use
+// the bootstrap ACL token as that's the only token available before creating an agent token
+// this determines whether we need to re-set the agent token every time Consul agent is restarted
+func (c *cmd) isACLTokenPersistent(bootstrapACLToken *string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:64]: ok, i'll change it to pass by value

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+			return emptyToken, fmt.Errorf("Failed to insert a new EdgeX agent token: %v", err)
+		}
+	}
+	return agentToken, nil
+}
+
+// getEdgeXAgentToken lists tokens and find the matched agent token by the expected key pattern
+// it returns the first matched agent token if many tokens actually are matched
+// it returns empty token if no matching found
+func (c *cmd) getEdgeXAgentToken(bootstrapACLToken *BootStrapACLTokenInfo) (string, error) {
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:170]: the `insertNewAgentToken` func also requires Policies from the token object so cannot just secretID alone

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		if err := json.NewDecoder(bytes.NewReader(createTokenResp)).Decode(&parsedTokenResponse); err != nil {
+			return emptyToken, fmt.Errorf("Failed to decode create token response body: %v", err)
+		}
+
+		c.loggingClient.Info("successfully created a new agent token")
+
+		return fmt.Sprintf("%s", parsedTokenResponse["SecretID"]), nil
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:347]: i am fine either way

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+		}
+	}
+	return agentToken, nil
+}
+
+// getEdgeXAgentToken lists tokens and find the matched agent token by the expected key pattern
+// it returns the first matched agent token if many tokens actually are matched
+// it returns empty token if no matching found
+func (c *cmd) getEdgeXAgentToken(bootstrapACLToken *BootStrapACLTokenInfo) (string, error) {
+	listTokensURL, err := c.getRegistryApiUrl(consulListTokensAPI)
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:171]: the check is not needed as this func only called by `createAgentToken` and that one has already checked; don't need to do the redundant checks

File: internal/security/bootstrapper/command/setupacl/acltokens.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:347]: actually, your suggestion not working: `cannot convert parsedTokenResponse["SecretID"] (map index expression of type interface{}) to string compiler`

[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:347]: I'll keep the original way I had

File: internal/security/bootstrapper/command/setupacl/command.go (hunk 1 above)
Comment (lenny-intel) [reply internal/security/bootstrapper/command/setupacl/command.go:191]: Actual, this doesn't work. 
The pointer itself is passed by value, so changing what it points to in the function does change the callers copy of the pointer. Best approach is to add `BootStrapACLTokenInfo` as a return value. Then the caller can use the returned value .

File: internal/security/bootstrapper/command/setupacl/acltokens.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go]: fixed

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+			resp.StatusCode, string(checkAgentResp))
+	}
+}
+
+// createAgentToken uses bootstrapped ACL token to create an agent token
+// this call requires ACL read/write permission and hence we use the bootstrap ACL token
+// it checks whether there is an agent token already existing and re-uses it if so
+// otherwise creates a new agent token
+func (c *cmd) createAgentToken(bootstrapACLToken *BootStrapACLTokenInfo) (string, error) {
+	if bootstrapACLToken == nil {
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go]: fixed

File: internal/security/bootstrapper/command/setupacl/acltokens.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go]: fixed

File: internal/security/bootstrapper/command/setupacl/command.go (hunk 1 above)
Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/command.go]: i am not changing what it points in this func. I have a new variable defined.

Comment (lenny-intel) [reply internal/security/bootstrapper/command/setupacl/command.go]: Right, just want to make sure. Does the caller need to new `BootStrapACLTokenInfo ` in this case?

Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/command.go]: no need, if the inputToken is nil (not new'ed), then it reconstructs it from the saved file- i've added the comments in this func.

**Review by lenny-intel — COMMENTED** [review]
LGTM defer to @bnevis-i for final approval

**Comment by bnevis-i** [comment]
@lenny-intel 
> LGTM defer to @bnevis-i for final approval

You still have a disapproving review filed.

**Comment by jim-wang-intel** [comment]
> LGTM defer to @bnevis-i for final approval

Don't worry, snap CI pipeline build failed anyway :-)  but the failure is nothing to do with this PR at all...

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+	consulCreateTokenAPI   = "/v1/acl/token"
+	// RUD: Read Update Delete
+	consulTokenRUDAPI = "/v1/acl/token/%s"
+)
+
+// isACLTokenPersistent checks Consul agent's configuration property for EnablePersistence of ACLTokens
+// it returns true if the token persistence is enabled; false otherwise
+// once ACL rules are enforced, this call requires at least agent read permission and hence we use
+// the bootstrap ACL token as that's the only token available before creating an agent token
+// this determines whether we need to re-set the agent token every time Consul agent is restarted
```
Comment (bnevis-i) [inline internal/security/bootstrapper/command/setupacl/acltokens.go:63]: Can't we just force EnablePersistence = true by setting in the config file?  Or if we need to check it, just outright fail if the precondition is not met?

Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:63]: we could, but i think this way is more robust. as I learned you can dynamically change conusl configure on the fly and reload.

Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:63]: https://www.consul.io/commands/reload

File: internal/security/bootstrapper/command/setupacl/acltokens.go [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+	 * For the details for each type, see reference https://www.consul.io/commands/acl/set-agent-token#token-types
+	 */
+	// DefaultType is agent token type "default" to be set
+	DefaultType AgentTokenType = "default"
+	// AgentType is agent token type "agent" to be set
+	AgentType AgentTokenType = "agent"
+	// MasterType is agent token type "master" to be set
+	MasterType AgentTokenType = "master"
+	// ReplicationType is agent token type "replication" to be set
+	ReplicationType AgentTokenType = "replication"
```
Comment (bnevis-i) [inline internal/security/bootstrapper/command/setupacl/acltokens.go:47]: Why do we need all of these defined if we only ever use AgentType in the code?

Comment (jim-wang-intel) [reply internal/security/bootstrapper/command/setupacl/acltokens.go:47]: don't need to just for the sake of completeness

**Review by bnevis-i — APPROVED** [review]

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by lenny-intel** [comment]
recheck

**Comment by lenny-intel** [comment]
> snap CI pipeline build failed anyway

That was a network error. rechecked.

**Pull request merged**